inc/namespace.php

```diff
@@ -977,34 +988,121 @@ function elasticpress_mapping( array $mapping ) : array {
 		// Ensure each field is represented in the mapping to avoid generating the
 		// default mapping.
 		foreach ( $fields as $field ) {
-			if ( ! isset( $mapping['mappings'][ $type ]['properties'][ $field ] ) ) {
-				$mapping['mappings'][ $type ]['properties'][ $field ] = [
-					'type' => 'text',
-				];
-			}
-			if ( ! isset( $mapping['mappings'][ $type ]['properties'][ $field ]['fields'] ) ) {
-				$mapping['mappings'][ $type ]['properties'][ $field ]['fields'] = [
-					'keyword' => [
-						'type' => 'keyword',
-						'ignore_above' => 256,
+			$field_mapping = [
+				'type' => 'text',
+				'fields' => [
+					'suggest' => [
+						'type' => 'text',
+						'analyzer' => 'edge_ngram_analyzer',
+						'search_analyzer' => $search_analyzer,
 					],
-					'raw' => [
-						'type' => 'keyword',
-						'ignore_above' => 256,
+				],
+			];
+
+			// Split field on dots in case this is a nested property.
+			$sub_fields = explode( '.', $field );
+
+			// This is a sub property so try and match against dynamic templates to get the mapping.
+			// Dynamic templates using match & unmatch operate on the last field in a given path.
+			$start_field = $sub_fields[0];
+			$end_field = $sub_fields[ count( $sub_fields ) - 1 ];
+			foreach ( ( $mapping['mappings'][ $type ]['dynamic_templates'] ?? [] ) as $template ) {
+				$template = wp_parse_args( $template[ array_keys( $template )[0] ], [
+					'match_pattern' => null,
+					'match' => '',
+					'unmatch' => '',
+					'path_match' => '',
+					'path_unmatch' => '',
+				] );
+				if ( $template['match_pattern'] === 'regex' && ! preg_match( "/{$template['match']}/", $end_field ) ) {
```

Comment:
These would benefit from some inline comments to explain what's happening in each of these breaks.

🔢